FreeRTOS/FreeRTOS-Kernel
# Fix xTaskNotifyWait & ulTaskNotifyTake determinism.

Description
-----------
Fixes the bug described in this issue:
https://github.com/FreeRTOS/FreeRTOS-Kernel/issues/612

One unorthodox thing being done here, which I feel should be pointed out, is that I am calling `vTaskSuspendAll()` from inside a critical section. This is done to minimize the performance impact of this change, since one of the main benefits of task notifications is that they are faster than the other primitives. The only worry I see here is that it may cause issues with certain ports if they do tricky things inside `portSOFTWARE_BARRIER()` or `portMEMORY_BARRIER()`, but I cannot think of any examples.

Test Steps
-----------
As suggested by Richard Barry, adding `configASSERT( ulCriticalNesting == 0UL )`, into `vListInsert()`, which is a condition that should always be met, will cause it to be triggered by a call to `xTaskNotifyWait()` or `ulTaskNotifyTake()` if the task blocks. This of course only works on ports that have the variable `ulCriticalNesting`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
